File: LeetCode/202. Happy Number/Solution.java
public class Solution {//use the same method as find circle in link list
    public boolean isHappy(int n) {
        int fast = getNext(getNext(n));
        int slow = getNext(n);
        while (fast != slow) {
            fast = getNext(getNext(fast));
            slow = getNext(slow);
        }
        return fast == 1;
    }

    public int getNext(int n) {
        int ans = 0;
        while (n > 0) {
            int digit = n % 10;
            ans += digit * digit;
            n /= 10;
        }
        return ans;
    }
}

File: LeetCode/202. Happy Number/solution.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """

        def process(t):
            ans = 0
            while t:
                ans += (t % 10) ** 2
                t /= 10
            return ans

        s = set()
        while n not in s:
            s.add(n)
            if n == 1:
                return True
            n = process(n)

        return False
